pkg/oidc/oidc.go

```diff
@@ -116,13 +119,47 @@ func (a *Authenticator) generateCodeVerifier(len int) (string, string) {
 	return codeVerifier, codeChallenge
 }
 
+func (a *Authenticator) renewToken(ctx context.Context, refreshToken string) (string, error) {
+	log.Info("renewing token by refresh token")
+	oauthToken := &oauth2.Token{
+		RefreshToken: refreshToken,
+	}
+	tokenSource := a.oauthConfig.TokenSource(ctx, oauthToken)
+	newToken, err := tokenSource.Token()
+	if err != nil {
+		return "", err
+	}
+
+	_, err = a.verifyToken(ctx, newToken)
+	if err != nil {
+		return "", err
+	}
+	return newToken.AccessToken, nil
+}
+
+func (a *Authenticator) verifyToken(ctx context.Context, oauth2Token *oauth2.Token) (*oidc.IDToken, error) {
+	rawIDToken, ok := oauth2Token.Extra("id_token").(string)
+	if !ok {
+		return nil, errors.New("No id_token field in oauth2 token")
+	}
+	idToken, err := a.tokenVerifier.Verify(ctx, rawIDToken)
+	if err != nil {
+		return nil, err
+	}
+	err = idToken.VerifyAccessToken(oauth2Token.AccessToken)
+	if err != nil {
+		return nil, err
+	}
+	return idToken, nil
+}
+
 // Authenticate with IDP
-func (a *Authenticator) Authenticate(ctx context.Context) (payload string, err error) {
+func (a *Authenticator) Authenticate(ctx context.Context, refreshToken string) (string, string, error) {
```

Comment:
It would be good to use type aliases or something rather than `(string, string, error)`